session_test.go

```diff
@@ -0,0 +1,313 @@
+// +build windows
+
+package etw
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/Microsoft/go-winio/pkg/etw"
+	"github.com/stretchr/testify/suite"
+	"golang.org/x/sys/windows"
+)
+
+type testProvider struct {
+	suite.Suite
+
+	provider *etw.Provider
+	guid     windows.GUID
+}
+
+func (p *testProvider) SetupSuite() {
+	provider, err := etw.NewProvider("TestProvider", nil)
+	p.Require().NoError(err, "Failed to initialize test provider.")
+	p.provider = provider
+	p.guid = (windows.GUID)(provider.ID)
+}
+
+func (p *testProvider) TearDownSuite() {
+	p.Require().NoError(p.provider.Close(), "Failed to close test provider.")
+}
+
+func (p testProvider) generateEvents(ctx context.Context, levels []etw.Level, fields ...etw.FieldOpt) {
+	if fields == nil {
+		fields = etw.WithFields(etw.StringField("TestField", "Foo"))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			for _, l := range levels {
+				_ = p.provider.WriteEvent("TestEvent", etw.WithEventOpts(etw.WithLevel(l)), fields)
+			}
+		}
+	}
+}
```

Comment:
Could we just generate a single event synchronously on demand? 

Seems it would be easier and more reliable to use in tests.

Reply:
Event processing starts with a little delay. When event processing is ready to consume events, we have already sent the event.

That's why I generate a lot of events in a goroutine.